### PR TITLE
TST: Adding merge test for non-string columns [Ref 17962]

### LINKED ISSUE #17962: BUG: merging with non-stringlike columns can raise
```
In [1]: pd.__version__
Out[1]: '0.21.0rc1+36.ge1dabf3'

In [2]: left = pd.DataFrame({
   ...:     0: [1, 0, 1, 0],
   ...:     1: [0, 1, 0, 0],
   ...:     2: [0, 0, 2, 0],
   ...:     3: [1, 0, 0, 0]
   ...: })
   ...: left
   ...: 
Out[2]: 
   0  1  2  3
0  1  0  0  1
1  0  1  0  0
2  1  0  2  0
3  0  0  0  0

In [3]: right = left.astype('float64')
   ...: right
   ...: 
Out[3]: 
     0    1    2    3
0  1.0  0.0  0.0  1.0
1  0.0  1.0  0.0  0.0
2  1.0  0.0  2.0  0.0
3  0.0  0.0  0.0  0.0

In [4]: pd.merge(left, right)
In [4]: pd.merge(left, right)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-ff1e14288b09> in <module>()
----> 1 pd.merge(left, right)

~/pandas-dev/pandas/core/reshape/merge.py in merge(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)
     55                          right_index=right_index, sort=sort, suffixes=suffixes,
     56                          copy=copy, indicator=indicator,
---> 57                          validate=validate)
     58     return op.get_result()
     59 

~/pandas-dev/pandas/core/reshape/merge.py in __init__(self, left, right, how, on, left_on, right_on, axis, left_index, right_index, sort, suffixes, copy, indicator, validate)
    567         # validate the merge keys dtypes. We may need to coerce
    568         # to avoid incompat dtypes
--> 569         self._maybe_coerce_merge_keys()
    570 
    571         # If argument passed to validate,

~/pandas-dev/pandas/core/reshape/merge.py in _maybe_coerce_merge_keys(self)
    928                 typ = lk.categories.dtype if lk_is_cat else object
    929                 self.left = self.left.assign(
--> 930                     **{name: self.left[name].astype(typ)})
    931             if name in self.right.columns:
    932                 typ = rk.categories.dtype if rk_is_cat else object

TypeError: assign() keywords must be strings

```

### PR DESCRIPTION
**Pull request by sofiane87**
- [x] closes #17962
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/tests/reshape/merge/test_merge.py [hunk 1]
```diff
@@ -134,6 +134,18 @@ def test_merge_common(self):
...
+        # https://github.com/pandas-dev/pandas/issues/17962
+        # Checks that method runs for non string column names
+        left = pd.DataFrame(
+            {0: [1, 0, 1, 0], 1: [0, 1, 0, 0], 2: [0, 0, 2, 0], 3: [1, 0, 0, 3]}
+        )
+
+        right = left.astype(float)
+        expected = left
+
+        tm.assert_frame_equal(expected, pd.merge(left, right))
```
Comment (mroeschke) [inline pandas/tests/reshape/merge/test_merge.py]: Could you assign `result = pd.merge(left, right)` before this statement so this line can be `tm.assert_frame_equal(expected, result)

**Comment by jreback** [comment]
lgtm, pending @mroeschke comment

File: pandas/tests/reshape/merge/test_merge.py (hunk 1 above)
Comment (sofiane87) [reply pandas/tests/reshape/merge/test_merge.py]: this is now done

**Comment by mroeschke** [comment]
Thanks @sofiane87!

**Pull request merged**